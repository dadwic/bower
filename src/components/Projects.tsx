import React from "react";
import Humanize from "humanize-plus";
import { useSearchParams } from "react-router-dom";
import Box from "@mui/system/Box";
import Card from "@mui/material/Card";
import CardHeader from "@mui/material/CardHeader";
import CardContent from "@mui/material/CardContent";
import IconButton from "@mui/material/IconButton";
import Typography from "@mui/material/Typography";
import Pagination from "@mui/material/Pagination";
import Skeleton from "@mui/material/Skeleton";
import Avatar from "@mui/material/Avatar";
import Rating from "@mui/material/Rating";
import Stack from "@mui/material/Stack";
import Link from "@mui/material/Link";
import TimeIcon from "@mui/icons-material/AccessTime";
import OpenInNewIcon from "@mui/icons-material/OpenInNew";
import { useAppSelector } from "hooks";
import { stringAvatar } from "utils";
import {
  selectLoading,
  selectProjects,
  selectProjectsCount,
} from "redux/selectors";
import Maybe from "./Maybe";
import Sort from "./Sort";
import moment from "moment";

export default function Packages() {
  const loading = useAppSelector(selectLoading);
  const projects = useAppSelector(selectProjects);
  const count = useAppSelector(selectProjectsCount);
  let [searchParams, setSearchParams] = useSearchParams();
  let searchPage: number = Number(searchParams.get("page") || 1);

  const handleChangePage = (_: React.ChangeEvent<unknown>, value: number) => {
    setSearchParams({
      ...Object.fromEntries(searchParams.entries()),
      page: value.toString(),
    });
  };

  return (
    <>
      <Maybe condition={loading}>
        <React.Fragment>
          <Skeleton
            sx={{
              width: { sm: "100%", md: "40%" },
              height: "calc(1.5em + 22px)",
            }}
          />
          {[...new Array(9)].map((_, key) => (
            <Card sx={{ mt: 2 }} key={key}>
              <CardHeader
                avatar={
                  <Skeleton
                    animation="wave"
                    variant="circular"
                    width={40}
                    height={40}
                  />
                }
                title={
                  <Skeleton
                    animation="wave"
                    height={20}
                    width="40%"
                    style={{ marginBottom: 6 }}
                  />
                }
                subheader={
                  <Box sx={{ display: "flex", alignItems: "center" }}>
                    <Rating
                      name="no-value"
                      size="small"
                      value={null}
                      readOnly
                    />
                    <Skeleton
                      animation="wave"
                      height={10}
                      width="10%"
                      style={{ marginLeft: 8 }}
                    />
                  </Box>
                }
              />
              <CardContent>
                <React.Fragment>
                  <Skeleton
                    animation="wave"
                    height={10}
                    style={{ marginBottom: 6 }}
                  />
                  <Skeleton animation="wave" height={10} width="80%" />
                </React.Fragment>
              </CardContent>
            </Card>
          ))}
        </React.Fragment>
      </Maybe>
      <Maybe condition={!loading}>
        <Stack spacing={2}>
          <Sort />
          {projects.map((project, key) => (
            <Card key={key}>
              <CardHeader
                title={project.name}
                titleTypographyProps={{
                  title: project.name,
                }}
                avatar={
                  <Avatar
                    component={Link}
                    href={project.repository_url || "#"}
                    target="_blank"
                    {...stringAvatar(
                      (project.repository_url || "")
                        .split("/")
                        .splice(-2, 1)
                        .toString()
                    )}
                  />
                }
                action={
                  <IconButton
                    component="a"
                    href={project.repository_url || "#"}
                    target="_blank"
                  >
                    <OpenInNewIcon />
                  </IconButton>
                }
                subheader={
                  <Box
                    sx={{
                      display: "flex",
                      alignItems: "center",
                    }}
                  >
                    <Rating
                      size="small"
                      name="project-stars"
                      value={project?.stars}
                      readOnly
                    />
                    <Stack
                      spacing={2}
                      direction="row"
                      alignItems="center"
                      sx={{ ml: 2, fontSize: 13 }}
                    >
                      <span>{Humanize.compactInteger(project.stars, 1)}</span>
                      <TimeIcon fontSize="small" />
                      <span>
                        {moment(
                          project.latest_release_published_at,
                          "YYYY-MM-DD"
                        ).fromNow()}
                      </span>
                    </Stack>
                  </Box>
                }
              />
              <CardContent>
                <Typography
                  variant="body2"
                  color="text.secondary"
                  component="p"
                >
                  {project?.description}
                </Typography>
              </CardContent>
            </Card>
          ))}
          <Pagination
            count={count}
            page={searchPage}
            variant="outlined"
            disabled={loading}
            onChange={handleChangePage}
          />
        </Stack>
      </Maybe>
    </>
  );
}
