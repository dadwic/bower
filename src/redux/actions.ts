import axios, { AxiosResponse, Canceler } from "axios";
import { AppThunk } from "./store";
import { API_PER_PAGE, API_KEY } from "../constants";
import {
  projectsLoading,
  projectsLoaded,
  projectsFailure,
} from "./projectsSlice";
import { Project } from "types/Project";
import { SortType } from "types";

const CancelToken = axios.CancelToken;
let cancel: Canceler;

export const fetchAsync =
  (text: string, sort: SortType, page: number): AppThunk =>
  async (dispatch) => {
    try {
      dispatch(projectsLoading());
      const response = await axios.get<Array<Project>>(
        `https://libraries.io/api/search?q=${text}&sort=${sort}&page=${page}&per_page=${API_PER_PAGE}&api_key=${API_KEY}`,
        {
          cancelToken: new CancelToken(function executor(c) {
            cancel = c;
          }),
        }
      );
      dispatch(
        projectsLoaded({
          data: response.data,
          total: Number(response.headers?.total),
        })
      );
    } catch (error: any) {
      if (axios.isCancel(error)) {
        console.log("Request canceled", error.message);
      } else {
        if (error.response) {
          const status = (error.response as AxiosResponse).status;
          if (status === 429) {
            return dispatch(
              projectsFailure("429 Too Many Requests! please try again")
            );
          }
          if (status >= 400 && status < 500) {
            return dispatch(projectsFailure("400 Bad Request"));
          }
          if (status >= 500) {
            return dispatch(projectsFailure("500 Internal Server Error"));
          }
        } else if (error.request) {
          console.log(error.request);
        } else {
          dispatch(projectsFailure(`Error ${error.message}`));
        }
      }
      console.log(error.config);
    }
  };
