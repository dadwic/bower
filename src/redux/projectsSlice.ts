import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { Project, ProjectsState } from "types/Project";

export const initialState: ProjectsState = {
  count: 0,
  error: null,
  loading: false,
  projects: [],
};

export const projectsSlice = createSlice({
  name: "projects",
  initialState,
  reducers: {
    projectsLoading: (state) => {
      state.loading = true;
    },
    projectsLoaded: (
      state,
      action: PayloadAction<{ data: Project[]; total: number }>
    ) => {
      state.projects = action.payload.data;
      state.count = action.payload.total;
      state.loading = false;
    },
    projectsFailure: (state, action: PayloadAction<string>) => {
      state.count = 0;
      state.projects = [];
      state.loading = false;
      state.error = action.payload;
    },
  },
});

export const { projectsLoading, projectsLoaded, projectsFailure } =
  projectsSlice.actions;

export default projectsSlice.reducer;
