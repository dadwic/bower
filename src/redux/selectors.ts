import { RootState } from "./store";

export const selectLoading = (state: RootState) => state.projects.loading;
export const selectProjects = (state: RootState) => state.projects.projects;
export const selectProjectsCount = (state: RootState) => state.projects.count;
export const selectProjectsError = (state: RootState) => state.projects.error;
