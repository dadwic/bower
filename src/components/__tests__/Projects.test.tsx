import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import { store } from "redux/store";
import Projects from "../Projects";

describe("Projects", () => {
  it("should corretly match snapshot", () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <BrowserRouter>
            <Projects />
          </BrowserRouter>
        </Provider>
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
