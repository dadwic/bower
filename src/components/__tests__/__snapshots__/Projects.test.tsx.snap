// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Projects should corretly match snapshot 1`] = `
<div
  className="css-1p5q5e5-MuiStack-root"
>
  <div
    className="css-1yo1jo6-MuiStack-root"
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
      data-testid="SortIcon"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M3 18h6v-2H3v2zM3 6v2h18V6H3zm0 7h12v-2H3v2z"
      />
    </svg>
    <button
      aria-expanded={false}
      aria-haspopup="listbox"
      className="MuiSelectUnstyled-root css-sv5l2l"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      tabIndex={0}
    >
      
    </button>
    <button
      className="MuiButton-root MuiButton-text MuiButton-textInherit MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorInherit MuiButtonBase-root  css-u44ztt-MuiButtonBase-root-MuiButton-root"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      Reset
    </button>
  </div>
  <nav
    aria-label="pagination navigation"
    className="MuiPagination-root MuiPagination-outlined css-1oj2twp-MuiPagination-root"
  >
    <ul
      className="MuiPagination-ul css-wjh20t-MuiPagination-ul"
    >
      <li>
        <button
          aria-label="Go to previous page"
          className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-lqq3n7-MuiButtonBase-root-MuiPaginationItem-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
            data-testid="NavigateBeforeIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
            />
          </svg>
        </button>
      </li>
      <li>
        <button
          aria-label="Go to next page"
          className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-outlined MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-lqq3n7-MuiButtonBase-root-MuiPaginationItem-root"
          disabled={true}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={-1}
          type="button"
        >
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
            data-testid="NavigateNextIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
            />
          </svg>
        </button>
      </li>
    </ul>
  </nav>
</div>
`;
